compiler/rustc_middle/src/ty/fold.rs

```diff
@@ -183,6 +183,12 @@ pub trait TypeFolder<'tcx>: Sized {
 
 pub trait TypeVisitor<'tcx>: Sized {
     type BreakTy = !;
+    /// Supplies the `tcx` for an unevaluated anonymous constant in case its default substs
+    /// are not yet supplied.
+    ///
+    /// Visitors which do not look into these substs may leave this unimplemented, so be
+    /// careful when calling this method elsewhere.
+    fn tcx_for_anon_const_substs<'a>(&'a self) -> TyCtxt<'tcx>;
```

Comment:
Is there a significant downside to just making this `tcx(&self)` and always requiring it? Is there any place we have a `TypeVisitor` but no tcx?

I also think that it's perhaps a bit nice to just have this be `tcx` and return an `Option<TyCtxt<'tcx>>`, then were it's used, we can just `expect`. But given the option, I think it's better to just always require a `TyCtxt`

Reply:
the reason I went for `tcx_for_anon_const_substs` is that I don't want to modify all the type flags stuff, because that would be a far bigger change.

Using `Option<TyCtxt<'tcx>>` seems ok though I would probably still keep the long name :thinking:
Using `None` incorrectly would cause some fairly bad issues in the future

Reply:
Why would you have to modify the type flags stuff to just add a function? Also, where are you expecting to use this function?

Reply:
i want to use this function to get the default anon const substs once these are lazily computed.

making `tcx` obligatory means that we have to modify stuff like `needs_subst` to take a `TyCtxt`, which is a more invasive change and shouldn't be needed.